### PR TITLE
[TEP046] Remove Deprecated PipelineRun Timeout Field

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes
This commit removes the `pipelineRun.spec.timeout` field which has been deprecated.  It tombstones the field and add the invalid error message for the validation if specified by users since it has just been the earliest date for removal.

It conforms to [api compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) and follows the timeline according to [deprecation.md](https://github.com/tektoncd/pipeline/blob/main/docs/deprecations.md).

/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: please migrate off `pipelineRun.spec.timeout` since it has been deprecated and is removed in v0.46
please use `pipelineRun.spec.timeouts.pipeline` instead
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/pipelinerun_defaults.go [hunk 1]
```diff
@@ -36,6 +36,10 @@ func (pr *PipelineRun) SetDefaults(ctx context.Context) {
 // SetDefaults implements apis.Defaultable
 func (prs *PipelineRunSpec) SetDefaults(ctx context.Context) {
 	cfg := config.FromContextOrDefaults(ctx)
+	if prs.Timeouts == nil {
+		prs.Timeouts = &TimeoutFields{}
+		prs.Timeouts.Pipeline = &metav1.Duration{Duration: time.Duration(cfg.Defaults.DefaultTimeoutMinutes) * time.Minute}
+	}
```
Comment (jerop) [inline pkg/apis/pipeline/v1/pipelinerun_defaults.go:42]: seems like this is a separate change to pipelinerun.timeouts -- could this be in a separate commit?

Comment (jerop) [reply pkg/apis/pipeline/v1/pipelinerun_defaults.go:42]: also we set prs.Timeouts.Pipeline twice - in line 41 and line 44

Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipelinerun_defaults.go:42]: Thanks Jerop. Yeah I will move it into a separate PR.

But to confirm on this addition, I might be mistaken but I wondered if the added condition would make sense if there are not any Timeouts specified. 

ie. previously the timeout default is set with PipelineRun.Timeout but it was not moved to v1 since timeout is deprecated.

**Comment by lbernick** [comment]
Is it necessary to remove support for pipelinerun.spec.timeout? It seems easy enough to just use pipelineRun.spec.timeouts.pipeline instead, which should be handled by existing conversion logic

File: pkg/apis/pipeline/v1/pipelinerun_defaults.go (hunk 1 above)
Comment (jerop) [reply pkg/apis/pipeline/v1/pipelinerun_defaults.go:42]: maybe this is what we need (replacing line 39 to line 45)
```go
	if prs.Timeouts == nil {
		prs.Timeouts = &TimeoutFields{}
	}
	if prs.Timeouts.Pipeline == nil {
		prs.Timeouts.Pipeline = &metav1.Duration{Duration: time.Duration(cfg.Defaults.DefaultTimeoutMinutes) * time.Minute}
	}
```

**Comment by JeromeJu** [comment]
> Is it necessary to remove support for pipelinerun.spec.timeout? It seems easy enough to just use pipelineRun.spec.timeouts.pipeline instead, which should be handled by existing conversion logic

Thanks Lee. Looking back on the field itself should not be a blocker for v1 storage version swap, but it could be better for cleaning up removable fields and  reconciler tests ie. pkg/reconciler/pipelinerun/pipeilnerun_tests

File: pkg/apis/pipeline/v1/pipelinerun_defaults.go (hunk 1 above)
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/pipelinerun_defaults.go]: Thanks Jerop, I edited it at https://github.com/tektoncd/pipeline/pull/6311

**Comment by lbernick** [comment]
> > Is it necessary to remove support for pipelinerun.spec.timeout? It seems easy enough to just use pipelineRun.spec.timeouts.pipeline instead, which should be handled by existing conversion logic
> 
> Thanks Lee. Looking back on the field itself should not be a blocker for v1 storage version swap, but it could be better for cleaning up removable fields and reconciler tests ie. pkg/reconciler/pipelinerun/pipeilnerun_tests

I think it would make sense to swap most of our tests to use pipelinerun.spec.timeouts.pipeline, and then have one or two tests reconciling a v1beta1 PipelineRun with pipelinerun.spec.timeout. WDYT?

**Comment by afrittoli** [comment]
@JeromeJu the unit test failure looks legit - could you check it?

**Comment by JeromeJu** [comment]
> @JeromeJu the unit test failure looks legit - could you check it?

thanks @afrittoli it was not removed until I updated with tombstoning:)

**Comment by pritidesai** [comment]
This PR did not make it to the 0.46.

@JeromeJu please rebase this, moving it to 0.47 for now.

**Comment by JeromeJu** [comment]
Closing since discussed offline with @lbernick this shall not be blocking v1 storage swap.
We could reopen this once timing is appropriate.
